hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieDataBlock.java

```diff
@@ -115,6 +114,35 @@ public byte[] getContentBytes() throws IOException {
     return serializeRecords(records.get());
   }
 
+  private Schema getReaderSchema(Option<Schema> readerSchemaOpt) {
+    Schema writerSchema = getWriterSchema(super.getLogBlockHeader());
+    // If no reader-schema has been provided assume writer-schema as one
+    if (!readerSchemaOpt.isPresent()) {
+      return writerSchema;
+    }
+
+    // Handle table renames when there are still log files
+    Schema readerSchema = readerSchemaOpt.get();
+    if (isHandleDifferingNamespaceRequired(readerSchema, writerSchema)) {
+      return writerSchema;
+    } else {
+      return readerSchema;
+    }
+  }
+
+  /**
+   * Spark3.1 uses avro:1.8.2, which matches fields by their fully qualified name. If namespaces are differs, reads will fail for fields that have the same name and type, but differing name(spaces).
+   * Such cases can arise when an ALTER-TABLE-RENAME ddl is performed.
+   *
+   * @param readerSchema the reader schema
+   * @param writerSchema the writer schema
+   * @return boolean if handling of differing namespaces between reader and writer schema are required
+   */
+  private static boolean isHandleDifferingNamespaceRequired(Schema readerSchema, Schema writerSchema) {
+    return readerSchema.getClass().getPackage().getImplementationVersion().compareTo("1.8.2") <= 0
+        && !readerSchema.getName().equals(writerSchema.getName());
+  }
```

Comment:
The fix works.  But I'm thinking at this layer of reading log files, such details should not be exposed.  It's better to fix the schema generation in ALTER-TABLE-RENAME DDL to have consistent namespaces, or resolve the schema's namespace on upper layers, e.g., `TableSchemaResolver`.  And the schema namespace should not change across Hudi commits.

Reply:
Agreed, the fix here was written with  backwards compatibility in mind. For tables that already have log files that were written in a certain format + namespace and have yet to be compacted, it is not realistic to modify those name spaces block by block. 

Given that this is an avro internal issue for lower avro version, and will only happen when performing merge. I thought it will be appropriate to put it here while reading log files.

There are a few ways to fix this like you mentioned and they are:

1. `ALTER-TABLE-RENAME-DDL` will rename the hive table name, but will not perform any internal renames. i.e. `hoodie.properties` will not be re-written, schema will hence be consistent.
2. Block `ALTER-TABLE-RENAME-DDL` if there are uncompacted log files; ALTER-TABLE-RENAME-DDL will be IO intensive, need to scan through all latest file slices to see if there are any uncompacted filegroups. log file list > 0 (akin to running a plan-compaction execution, and if compaction operations is not empty, we will reject the rename)
3. Fix it while reading as seen in the fix here (code is ugly and we are coupling business logic + with dependency version)